File: scripts/import_codes.js
import { readFileSync } from 'fs';
import csvParser from 'csv-parser';
import { Readable } from 'stream';
import { db } from '../server/db.ts';
import { codes } from '../shared/schema.ts';
import 'dotenv/config';

async function importCodes() {
  console.log('Starting BD Code RAMQ import...');
  
  // Read CSV file
  const csvData = readFileSync('data/imports/RAMQ-codes.csv', 'utf8');
  
  const records = [];
  
  // Parse CSV
  const readable = Readable.from([csvData]);
  
  return new Promise((resolve, reject) => {
    readable
      .pipe(csvParser())
      .on('data', (row) => {
        // Skip rows with empty billing_code
        if (!row.billing_code || row.billing_code.trim() === '') {
          console.log('Skipping row with empty billing_code:', Object.keys(row));
          return;
        }

        // Map CSV fields to database schema
        const codeRecord = {
          code: row.billing_code.trim(),
          description: row.description || '',
          category: row.level1_group || '',
          place: row.place || '',
          tariffValue: row.tariff_value ? parseFloat(row.tariff_value) : null,
          extraUnitValue: row.extra_unit_value ? parseFloat(row.extra_unit_value) : null,
          unitRequire: row.unit_require === 'TRUE',
          sourceFile: row.source_file || '',
          topLevel: row.top_level || '',
          level1Group: row.level1_group || '',
          level2Group: row.level2_group || '',
          leaf: row.leaf || '',
          indicators: row.indicators || '',
          anchorId: row.anchor_id || '',
          active: true,
          customFields: {},
          updatedAt: new Date(),
          updatedBy: 'system_import'
        };

        records.push(codeRecord);
      })
      .on('end', async () => {
        try {
          console.log(`Parsed ${records.length} code records`);
          
          // Insert records in batches to avoid memory issues
          const batchSize = 100;
          let inserted = 0;
          
          for (let i = 0; i < records.length; i += batchSize) {
            const batch = records.slice(i, i + batchSize);
            
            try {
              await db.insert(codes).values(batch);
              inserted += batch.length;
              console.log(`Inserted batch ${Math.floor(i/batchSize) + 1}, total: ${inserted}`);
            } catch (error) {
              console.error(`Error inserting batch:`, error);
              // Continue with next batch
            }
          }
          
          console.log(`✅ Import completed! Inserted ${inserted} code records`);
          resolve();
        } catch (error) {
          console.error('Error during import:', error);
          reject(error);
        }
      })
      .on('error', reject);
  });
}

// Run the import
importCodes()
  .then(() => {
    console.log('Import finished successfully');
    process.exit(0);
  })
  .catch((error) => {
    console.error('Import failed:', error);
    process.exit(1);
  });

File: scripts/import_codes.py
#!/usr/bin/env python3

import csv
import json
import psycopg2
import os
from urllib.parse import urlparse

# Parse DATABASE_URL
DATABASE_URL = os.environ['DATABASE_URL']
url = urlparse(DATABASE_URL)

# Connect to database
conn = psycopg2.connect(
    host=url.hostname,
    port=url.port,
    user=url.username,
    password=url.password,
    database=url.path[1:]  # Remove leading slash
)
cur = conn.cursor()

print("Starting BD Code RAMQ import...")

# Read and process CSV
with open('attached_assets/BD Code RAMQ - ramq_all (1)_1758943164183.csv', 'r', encoding='utf-8') as file:
    reader = csv.DictReader(file)
    
    batch = []
    count = 0
    
    for row in reader:
        # Create composite code: billing_code + place (for uniqueness)
        billing_code = row['billing_code']
        place = row['place'] or 'all'
        
        # Use the billing_code with place suffix for uniqueness
        if place == 'all':
            code = billing_code
        else:
            code = f"{billing_code}-{place}"
        
        # Map CSV fields to database schema
        record = {
            'code': code,
            'description': row['description'] or '',
            'category': place,
            'active': row['unit_require'] == 'FALSE',  # FALSE in CSV = active
            'custom_fields': json.dumps({
                'billing_code': billing_code,
                'place': place,
                'tariff_value': row['tariff_value'],
                'extra_unit_value': row['extra_unit_value'],
                'source_file': row['source_file'],
                'top_level': row['top_level'],
                'level1_group': row['level1_group'],
                'level2_group': row['level2_group'],
                'leaf': row['leaf'],
                'indicators': row['indicators'],
                'anchor_id': row['anchor_id']
            }),
            'updated_by': 'system_import'
        }
        
        batch.append(record)
        
        # Insert in batches of 100
        if len(batch) >= 100:
            try:
                # Use execute_values for efficient batch insert
                from psycopg2.extras import execute_values
                
                execute_values(
                    cur,
                    """
                    INSERT INTO codes (code, description, category, active, custom_fields, updated_at, updated_by)
                    VALUES %s
                    ON CONFLICT (code) DO UPDATE SET
                        description = EXCLUDED.description,
                        category = EXCLUDED.category,
                        active = EXCLUDED.active,
                        custom_fields = EXCLUDED.custom_fields,
                        updated_at = EXCLUDED.updated_at,
                        updated_by = EXCLUDED.updated_by
                    """,
                    [(r['code'], r['description'], r['category'], r['active'], 
                      r['custom_fields'], 'NOW()', r['updated_by']) for r in batch],
                    template=None
                )
                
                count += len(batch)
                print(f"Inserted batch: {count} records")
                conn.commit()
                
            except Exception as e:
                print(f"Error inserting batch: {e}")
                conn.rollback()
            
            batch = []
    
    # Insert remaining records
    if batch:
        try:
            from psycopg2.extras import execute_values
            
            execute_values(
                cur,
                """
                INSERT INTO codes (code, description, category, active, custom_fields, updated_at, updated_by)
                VALUES %s
                ON CONFLICT (code) DO UPDATE SET
                    description = EXCLUDED.description,
                    category = EXCLUDED.category,
                    active = EXCLUDED.active,
                    custom_fields = EXCLUDED.custom_fields,
                    updated_at = EXCLUDED.updated_at,
                    updated_by = EXCLUDED.updated_by
                """,
                [(r['code'], r['description'], r['category'], r['active'], 
                  r['custom_fields'], 'NOW()', r['updated_by']) for r in batch],
                template=None
            )
            
            count += len(batch)
            conn.commit()
            
        except Exception as e:
            print(f"Error inserting final batch: {e}")
            conn.rollback()

print(f"✅ Import completed! Inserted {count} code records")

# Close connection
cur.close()
conn.close()